indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialDimensionDistributionParallelIndexTaskRunner.java

```diff
@@ -58,6 +84,188 @@ public String getName()
+   * temp directory.
+   * <p>
+   * This method is not thread-safe as the reports are processed by a single-threaded executor.
+   */
+  private void extractDistributionsFromReport(DimensionDistributionReport report)
+  {
+    log.debug("Started writing distributions from Task ID [%s]", report.getTaskId());
+
+    if (report.getIntervalToDistribution() == null) {
+      return;
+    }
+    report.getIntervalToDistribution().forEach(
+        (interval, distribution) -> {
+          Set<String> taskIds = intervalToTaskIds.computeIfAbsent(interval, i -> new HashSet<>());
+          final String subTaskId = report.getTaskId();
+          if (!taskIds.contains(subTaskId)) {
+            writeDistributionToFile(interval, subTaskId, distribution);
+            taskIds.add(subTaskId);
+          }
+        }
+    );
+
+    log.debug("Finished writing distributions from Task ID [%s]", report.getTaskId());
+  }
+
+  /**
+   * Writes the given distribution to the task temp directory.
+   * <p>
+   * If this operation fails, it requests a graceful shutdown of the runner via
+   * {@link #stopGracefully(String)}.
+   */
+  private void writeDistributionToFile(
+      Interval interval,
+      String subTaskId,
+      StringDistribution distribution
+  )
+  {
+    try {
+      File intervalDir = getIntervalDistributionDir(interval);
+      FileUtils.mkdirp(intervalDir);
+
+      File distributionJsonFile = getDistributionJsonFile(intervalDir, subTaskId);
+      getToolbox().getJsonMapper().writeValue(distributionJsonFile, distribution);
+    }
+    catch (IOException e) {
+      String errorMsg = StringUtils.format(
+          "Exception while writing distribution file for Interval [%s], Task ID [%s]",
+          interval,
+          subTaskId
+      );
+      stopGracefully(errorMsg);
+      throw new ISE(e, errorMsg);
+    }
+  }
+
+  private StringDistribution readDistributionFromFile(File intervalDir, String subTaskId)
+  {
+    try {
+      File distributionJsonFile = getDistributionJsonFile(intervalDir, subTaskId);
+      return getToolbox().getJsonMapper().readValue(distributionJsonFile, StringDistribution.class);
+    }
+    catch (IOException e) {
+      throw new ISE(e, "Error while reading distribution for Interval [%s], Task ID [%s]",
+                    intervalDir.getName(), subTaskId
+      );
+    }
+  }
+
+  private File getIntervalDistributionDir(Interval interval)
+  {
+    return new File(tempDistributionsDir, toIntervalString(interval));
+  }
+
+  private File getDistributionJsonFile(File intervalDir, String subTaskId)
+  {
+    return new File(intervalDir, subTaskId);
+  }
+
+  /**
+   * Waits for distributions from pending reports (if any) to be extracted.
+   */
+  private void waitToProcessPendingReports()
+  {
+    log.info("Waiting to extract distributions from sub-task reports.");
+    try {
+      executor.shutdown();
+      executor.awaitTermination(60, TimeUnit.SECONDS);
```

Comment:
if I just read this logic, I wouldn't read it as "oh, it's waiting for all task reports to happen" I'd read it as "this is a shutdown routine".  Given that the task knows how many sub tasks are happening, it might be nicer to have a countdown latch initialized with the number of subtasks that this waits on.

Reply:
Thanks for the suggestion! CountDownLatch would be much better.

Reply:
Maybe waiting for futures from submit should also be fine?

Reply:
Used a `Phaser` as it seemed to be the simplest for this use case.
Couldn't use a `CountdownLatch` as we only know the "estimated" number of "successful" tasks while starting the task runner. The actual number of tasks can vary and we might even receive reports for failed tasks (verified this during testing).

Using Futures would effectively achieve the same behaviour as using the Phaser i.e. await completion of all submitted tasks. But I feel it might have been a bit of an overkill as we really just need the count, not the individual futures.